---
title: "Week 4: Bonus Material"
author: 
- "Andrew Eckert (worked example)"
- "Helene Wagner (vignette)"
date: "`r Sys.Date()`"
show_toc: true
output:
  knitr:::html_vignette:
    toc: yes
    fig_width: 4 
    fig_height: 3.5
vignette: >
  %\VignetteIndexEntry{Week 4: Bonus Material}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## 1. Overview of Bonus Material

### a) Goals 

This bonus material shows how to:

- Test whether or not phenotypic trait differentiation is statistically different than genetic differentiation at random molecular genetic markers (SNPs).
- Test whether molecular markers behave neutrally ($F_{ST}$ outliers)
- Test association between trait and molecular markers

### b) Data set 

- **WWP_SNP_genotypes.txt**: SNP genotypes for all trees sampled in the field.
- **WWP_environmental_data.txt**:Environmental data collected from each plot within each population.
- **WWP_phenotype_data.txt**: Phenotypic measurements for 5 seedlings per tree made in a common garden.

### c) Required R libraries

All required packages should have been installed already when you installed 'LandGenCourse'.

```{r message=FALSE, warning=TRUE}
#devtools::install_github("kjgilbert/QstFstComp")
#source("http://bioconductor.org/biocLite.R")
#biocLite("LEA")
require(here)
require(hierfstat)
require(QstFstComp)
require(LEA)
source(paste0(here(),"/data/supplemental_R_functions.R"))
if(!dir.exists(paste0(here(),"/data"))) dir.create(paste0(here(),"/data"))
```

## 2. Estimate $F_{ST}$ from SNP data

**Motivation**: Now that we have shown that genetic variation for $δ^{13}C$ within populations is significantly greater than zero (i.e. $h^2 > 0$), that differentiation for $δ^{13}C$ is statistically greater than zero (i.e. $Q_{ST} > 0$), and that climate, and to a lesser degree geography, is correlated with $δ^{13}C$ values, we can formally test whether or not differentiation for $δ^{13}C$ is unexplainable due to neutral processes such as genetic drift and gene flow. The general idea is use a set of genetic markers we think primarily reflects neutral processes to estimate what $Q_{ST}$ should be without any form of natural selection operating in our system. To do that, we will use 164 SNPs sampled from gene regions that have no apparent functional connection to $δ^{13}C$. This will allow us to conclude that the differentiation we see is not just different from zero (done before), but different than expectations from a neutral model.

**Goals & Background**: The goal for this part of the laboratory is to test the hypothesis that $Q_{ST} > F_{ST}$. We will do that using SNP data that are stored in the file named "WWP_SNP_genotypes.txt". As with the previous files, this is a tab-delimited text file.

### a) Import and check SNP data 

First, we need to import the SNP data, and also recreate the phenotypic and environmental data (Week 4).

```{r}
snp <- read.delim(paste0(here(),"/data/WWP_SNP_genotypes.txt"), 
                  sep = "\t", header = T)
phen <- read.delim(paste0(here(),"/data/WWP_phenotype_data.txt"), 
                   sep = "\t", header = T)
env <- read.delim(paste0(here(),"/data/WWP_environmental_data.txt"), 
                  sep = "\t", header = T)
env2 <- data.frame(matrix(nrow=nrow(env), ncol=ncol(env)))
colnames(env2) <- colnames(env)
env2[,c(1:2)] <- env[,c(1:2)]
for(i in 3:ncol(env2)) {env2[,i] <- scale(env[,i], center = T, scale = T)}
```

Now, we need to convert the format of the SNP genotypes to FSTAT format for use in hierfstat. Please use the 'hierfstat_convert' function from those provided to you in the file named "supplemental_R_functions.R":

```{r}
snp_reformat <- hierfstat_convert(snp = snp, ids = c(1,2))
```

Let’s check for weird artifacts. Sometimes, for technical molecular biology reasons, a SNP in a data file turns out to be monomorphic (i.e. it has no variation). We can check this using:

```{r}
mono <- numeric(ncol(snp_reformat))
for (i in 1:ncol(snp_reformat)) 
{
  mono[i] <- length(table(snp_reformat[,i]))
}
snp_reformat2 <- snp_reformat[,-which(mono == 1)]
```

Now, we need to add names to the SNPs that are renaming and create the population identifiers for hierfstat. We can do this using the following:


```{r}
colnames(snp_reformat2) <- colnames(snp)[3:ncol(snp)][-which(mono == 1)]
pop_ids <- data.frame(matrix(snp[,2], nrow=nrow(snp), ncol = 1))
colnames(pop_ids) <- c("population")
```

### b) Estimate global $F_{ST}$ 

Let’s examine differentiation for the SNP data using $F_{ST}$. There are a multitude of ways to do this, but we will use the varcomp.glob function from the 'hierfstat' package:

```{r}
fst <- varcomp.glob(levels = pop_ids, loci = snp_reformat2, diploid = T)
```

Note you can also explore bootstrapping across loci to get a confidence interval using the boot.vc() function.

Now, let’s look at the output. The object fst has three elements. The first element is matrix of variance components for each SNP ('loc'). The columns of this matrix are levels you used from the highest to the lowest (left to right). For us, that means column 1 is the variance component for population, column 2 is the variance component for individual, and column 3 is the variance component for the error (or residual). The second element is the sum of the columns ('overall'). The last element is a matrix of F-statistics ('F'). These work by using as subscripts the column title relative to the row title, so the first value on the first line is the F- statistic for population relative to total (i.e. $F_{ST}$). It is calculated based on the variance components from $overall as:

$$F_{pop,tot} = \frac{\sigma^{2}_{pop}}
{\sigma^{2}_{pop}+\sigma^{2}_{ind}+\sigma^{2}_{error}}$$

### c) Compare $Q_{ST}$ to $F_{ST}$

Now that we have inspected overall genetic differentiation among populations, let’s use the QstFstComp library to formally test whether or not $Q_{ST} > F_{ST}$ for $δ^{13}C$.

```{r}
phen_mod <- phen[,-c(2,4)]
snp_reformat3 <- cbind(snp[,2], snp_reformat2); colnames(snp_reformat3)[1] <- c("population")
QstFst_out <- QstFstComp(fst.dat = snp_reformat3, qst.dat = phen_mod, 
                         numpops = 10, nsim = 10000, 
                         breeding.design = "half.sib.dam", 
                         dam.offspring.relatedness = 0.25, output = "concise")
QstFst_out
```

Inspect the first and third elements of the list 'QstFst_out'. Is $Q_{ST} > F_{ST}$? What does this mean biologically? Why is the estimated $Q_{ST}$ a little higher here as opposed to the point estimate from before?

## 3. Test for $F_{ST}$ outlier loci

### a) Estimate $F_{ST}$ for each SNP

Let’s inspect the results by SNP. We can calculate $F_{ST}$ for each SNP (or any locus for that matter) using the equation above and the variance components in $loc. I have provided a function to do this for you in the "supplemental_R_functions.R" file: 'fst_persnp'.

```{r}
fst_snp <- fst_persnp(vc = fst$loc, names = colnames(snp_reformat2))
```

Inspect the variation across loci relative to the global (multilocus value). Please realize that negative values should be considered as 0. These values are artifacts of estimating the variance components with finite sample sizes. 

### b) Estimate heterozygosity for each SNP

Use the 'het_snp' function in the "supplemental_R_functions.R" file to get heterozygosity. 
```{r}
het_out <- het_snp(snp=snp_reformat2, finite.cor= T, names = colnames(snp_reformat2))
```

Plot $F_{ST}$ against heterozygosity (left). To make a plot like those from Arlequin, you can divide the heterozygosity values by $(1 – F_{ST})$ prior to plotting them (right).

```{r, fig.show='hold'}
plot(het_out, fst_snp)
plot(het_out/(1-fst_snp), fst_snp)
```

### c) Test for $F_{ST}$ outlier loci

ADD THIS PART?

## 4. Test association between trait and genetic data

ADD THIS PART?




