---
title: "Week 4: Conceptual Exercise"
author: 
- "Andrew Eckert (worked example)"
- "Helene Wagner (vignette)"
date: "`r Sys.Date()`"
output:
  knitr:::html_vignette:
    fig_width: 4 
    fig_height: 3.5
vignette: >
  %\VignetteIndexEntry{Week 4: Conceptual Exercise}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

### a) Goals 

**Motivation**: Natural selection acts on phenotypic variation that is genetically determined. As such, it can be difficult to get a complete picture about adaptation from scanning genomes using molecular markers. The reason is that genetic outliers, even if true positives, have little to no information present about what phenotype they affect and how this phenotype results in fitness differences. Moreover, it is debatable to scan genomes for the presence of outliers if you have yet to demonstrate that the populations being sampled are locally adapted.

**Learning Objectives**: This assignment focuses on the interpretation of basic quantitative and population genetic analyses useful to testing hypotheses about local adaptation. Phenotypic measurement is undergoing a revolution, so that familiarity with basic methods in quantitative genetics will serve you well in the future. 

In addition, this week's worked example (computer lab) shows how to run the analyses to generate the output shown here.

### b) Data set 

The data come from a study of western white pine (*Pinus monticola* Dougl. ex D. Don) sampled around the Lake Tahoe Basin of California and Nevada. These data consist of 157 trees sampled from 10 populations (n = 9 to 22 trees/population). Within each population, trees were sampled within three plots. For each plot, GPS coordinates were collected (i.e. each plot in each population has its own GPS coordinates) and used to generate a set of 7 environmental variables. From these trees, needle tissue was collected from which total DNA was extracted and genotyped for 164 single nucleotide polymorphisms (SNPs). Seeds were also collected and planted in a common garden. The seedlings (n = 5 to 35/tree) were measured for several phenotypic traits. 

The phenotypic trait we will be working with today is known as the carbon isotope ratio ($δ^{13}C$). It is the ratio of two isotopes of carbon ($^{13}C$ and $^{12}C$) relative to an experimental control, and it is strongly correlated with intrinsic water-use efficiency in plants. Plants need water to live, so it is not a stretch of the imagination to believe that this phenotypic trait has a link with plant fitness.

We thus have three types of data:

- Phenotypic measurements for 5 seedlings per tree made in a common garden.
- Environmental data collected from each plot within each population.
- SNP genotypes for all 157 trees sampled in the field 

```{r message=FALSE, warning=TRUE, echo=FALSE}

#devtools::install_github("krlmlr/here")
require(here)
require(LandGenCourse)
require(lme4)
require(car)
#require(EcoGenetics)
require(tibble)
require(hierfstat)
require(QstFstComp)

source(paste0(here(),"/data/supplemental_R_functions.R"))

if(!dir.exists(paste0(here(),"/data"))) dir.create(paste0(here(),"/data"))
```

## Part 1: Estimate genetic values for an observed phenotypic trait

**Motivation**: A lot of genetics can be carried out without use of any molecular markers. Practitioners of empirical population genetics forget this quite often. A common garden allows us to create a standardized environment in which we minimize the influence of environment on the expression of a particular phenotypic trait. Since we know, after over a century of showing it to be true, that phenotypic variation results from genetic variation, environmental variation, and the interaction of genetic and environmental variation, then if we standardize the environment, phenotypic variation we see in the common garden is due to genetic variation (or if multiple gardens are used, genetics and the interaction of genetics and the environment).

**Goals**: The goal for this part is to model $δ^{13}C$ measurements of seedlings as a combined effect of family (mother tree) and population (where mothers are nested within populations). Also included is an experimental treatment known as “block”. In a common garden, seedlings from the same maternal tree are randomized among blocks to avoid the influence of micro-environmental variation on expression of phenotypic traits.

```{r, echo=FALSE}
phen <- read.delim(paste0(here(),"/data/WWP_phenotype_data.txt"), 
                   sep = "\t", header = T)
#as.tibble(phen)
```

We will compare three models (all models include a fixed effect of block): 

- **Model 1**: a fixed effect model with only an intercept, 
- **Model 2**: a LMM (linear mixed model) with an intercept (fixed) and a random effect due to family, and 
- **Model 3**: a LMM with an intercept, a random effect due to family nested within population, and a random effect of population. 

```{r, echo=TRUE}
mod1<- lm(phen$d13c~1+phen$block)
mod2<- lmer(d13c~1+(1|family)+block,data = phen, REML = F)
mod3 <- lmer(d13c ~ 1 + (1|population/family) + block, data = phen, REML = F)
```

**Question 1**: 
*Compare the output for the three models below. What do you conclude about the effect of block in each of the models? Remember this is an experimental treatment, so your conclusion directly addresses whether micro-environmental variation exists in your common garden.*

```{r, echo=FALSE}
cat("\n", "Model 1:", "\n")
Anova(mod1, type="III", test.statistic = "F")
cat("\n", "Model 2:", "\n")
Anova(mod2, type="III", test.statistic = "Chisq")
cat("\n", "Model 3:", "\n")
Anova(mod3, type="III", test.statistic = "Chisq")
```

Now, let’s explore which model best fits our data. To do this we will use the Akaike Information Criterion (AIC). This statistic is like a penalized likelihood score, where the penalty increases as the number of parameters in a model increases. When using AIC, the model with the lowest score is the preferred model. 

```{r, echo=FALSE}
aic_vals <- c(AIC(mod1), AIC(mod2), AIC(mod3))
names(aic_vals) <- c("mod1","mod2","mod3")
cat("\n", "AIC values:", "\n")
aic_vals
```
**Question 2**: 
*Which model fits the data best?*

We can express the relative support of each model using Akaike weights. They add to one and can be thought of as conditional probabilities of each model, with the conditioning being on only these three models being examined. 

```{r, echo=FALSE}
aic_out <- aic_weights(aic_vals)
cat("\n", "Akaike weights:", "\n")
aic_out
```

**Question 3**:

- *Which model has the highest probability?* 
- *How much larger is it than the other probabilities?* 
- *What does this tell you about your optimal model?*

```{r, echo=FALSE}
mod3_eff <- ranef(mod3)
mod3_fam_only <- mod3_eff$family + -30.59964
mod3_all_eff <- mod3_fam_only + pop_rep(pop.eff = mod3_eff$population, 
                                        n.fam = nrow(mod3_eff$family), 
                                        fam.eff = mod3_eff$family)
```

## Part 2: Estimate trait heritability

**Motivation**: Once we have estimated genetic values for $δ^{13}C$ with the best model from above, we can estimate what fraction of the total variation in trait values is due to genetic effects and how much of this genetic effect is due to families nested within populations and to populations. These analyses provide key information about whether or not local adaptation should even be considered. Remember that local adaptation is about genetically determined phenotypes that vary across environments in responses to differing selective pressures. This step allows us to assess how genetic variation for a phenotypic trait is distributed across the landscape.

**Goals**: The goal for this part is to estimate heritability, trait differentiation, and correlation with environment for trait values determined in Part 1. 

### a) Heritability

Let’s start with estimating the heritability of $δ^{13}C$. If you remember from your undergraduate evolution course, heritability refers generally to the proportion of phenotypic variance due to genetic variance. It comes in at least two different versions. The version we are interested in is narrow-sense heritability ($h^2$), which is defined as the ratio of additive genetic variance to total phenotypic variance:

$$h^{2} = \frac{\sigma^{2}_{additive}}{\sigma^{2}_{total}}$$

If we assume that the seedlings from each maternal tree are half-siblings (i.e. same mom, but each with a different father) then $σ^2_{additive} = 4 σ^2_{family}$ (so variance due to family:population). If the seedlings were all full-siblings, then the 4 would be replaced with 2. 

```{r, echo=FALSE}
add_var <- 4*(0.2831^2)
total_wp_var <- (0.2831^2) + (0.8509^2)
h2 <- add_var/total_wp_var
cat("\n", "Estimate of h^2:", "\n")
h2
```

**Question 4**: 
*What does this value mean?*

Here is a 95% bootstrap confidence interval for our estimate of $h^2 = 0.399$. The boxplot below shows the bootstrap distribution, i.e., the distribution of $h^2$ values that we might expect if the trait was not heritable.

```{r, echo=FALSE}
h2_boot_out <- mod_boot(model = mod3, nboot = 1000)
ci_95 <- quantile(h2_boot_out, probs = c(0.025, 0.50, 0.975)) ### 95% ci.
cat("\n", "95% bootstrap confidence interval:", "\n")
ci_95
boxplot(h2_boot_out, range=5); abline(h = h2, col = "red") 
## red line: original h2 estimate for comparison to bootstrap distribution.
```

Compare the numerical 95% confidence interval and boxplot with our original estimate of $h^2= 0.399$ (red line). 

**Question 5**:

- *Do you think that $h^2$ is statistically different than zero?* 
- *Is this consistent with the AIC results from Part 1?* 
- *Is it meaningful that the red line is very similar to the mean (or median) of the bootstrap distribution?* 

### b) Trait differentiation

Great, we have shown that within population genetic variation is statistically greater than zero. What about among population genetic variation? Let’s get to that right now. To measure among population genetic variation we will use a statistic known as $Q_{ST}$. It is similar in concept to $F_{ST}$ from population genetics. If we assume that all seedlings are again half-siblings, then:

$$Q_{ST} = \frac{\sigma^{2}_{population}}
{\sigma^{2}_{population}+8\sigma^{2}_{family}}$$

```{r, echo=FALSE}
num_qst <- 0.3088^2
dem_qst <- (0.3088^2) + (8*(0.2831^2))
qst <- num_qst/dem_qst
cat("\n", "Estimate of QST:", "\n")
qst
```

**Question 6**:

- *What does this value mean?* 
- *Look at the quantities in the equation above, what is the denominator equal to?* 
- *Is it the total phenotypic variance or the total genetic variance?*

Now, we can again look at a confidence interval using parametric bootstrapping. 

```{r, echo=FALSE}
qst_boot_out <- mod_boot_qst(model = mod3, nboot = 1000)
ci_95_qst <- quantile(qst_boot_out, probs = c(0.025, 0.50, 0.975)) ### 95% ci.
cat("\n", "95% bootstrap confidence interval:", "\n")
ci_95_qst
boxplot(qst_boot_out); abline(h = qst, col = "red")
```

**Question 7**: Interpret the results. 

- *Do you think that $Q_{ST}$ is statistically different than zero?* 
- *Is this consistent with the AIC results from Part 1?* 
- *Is it meaningful that the red line is less similar to the mean (or median) of the bootstrap distribution as compared to $h^2$?*

### c) Trait correlation with environment

Finally, let's test for correlations between genetic values of $δ^{13}C$ and environmental data. This output below summarizes three different regression models that model the trait $δ^{13}C$ as a function of a number of environmental variables. These include spatial coordinates (longitude, latitude), elevation, and several bioclimatic variables:

- **max_rad**: maximum solar radiation
- **tmax_july**: July maximum temperature
- **tmin_jan**: January minimum temperature
- **ann_ppt**: annual precipitation
- **gdd_aug**: growing degree days in August (i.e, temperature > 10C)
- **AWS050**: available water supply (soil property)

Here we fit three alternative models, accounting ('conditioning') for one group of variables while estimating the effect of another group of variables: 

- Model 1: effects of geography and climate
- Model 2: effect of climate, conditioned on geography
- Model 3: effect of geography, conditioned on climate

```{r, echo=FALSE}
env <- read.delim(paste0(here(),"/data/WWP_environmental_data.txt"), 
                  sep = "\t", header = T)
env2 <- data.frame(matrix(nrow=nrow(env), ncol=ncol(env)))
colnames(env2) <- colnames(env)
env2[,c(1:2)] <- env[,c(1:2)]
for(i in 3:ncol(env2)) {env2[,i] <- scale(env[,i], center = T, scale = T)}

phen_env <- cbind(mod3_all_eff[,1], env2[,c(3:11)])
colnames(phen_env) <- c("d13c", colnames(env2)[3:11])
mod1_env <- lm(d13c ~ longitude + latitude + elev + max_rad + tmax_july + 
                 tmin_jan + ann_ppt + gdd_aug + AWS050, data = phen_env)
cat("\n", "Model 1: geography and climate", "\n")
summary(mod1_env)
```

```{r, echo=FALSE}
res_geog <- residuals(lm(d13c ~ longitude + latitude, data = phen_env))
phen_env_mod2 <- cbind(res_geog, env2[,5:11])
mod2_env <- lm(d13c ~ elev + max_rad + tmax_july + tmin_jan + ann_ppt + gdd_aug + AWS050, data = phen_env)
cat("\n", "Model 2: climate, conditioned on geography", "\n")
summary(mod2_env)
```

```{r, echo=FALSE}
res_clim <- residuals(lm(d13c ~ elev + max_rad + tmax_july + tmin_jan + ann_ppt + gdd_aug + AWS050, data = phen_env))
phen_env_mod3 <- cbind(res_clim, env2[,c(3:4)]); colnames(phen_env_mod3) <- c("d13c", "longitude", "latitude")
mod3_env <- lm(d13c ~ longitude + latitude, data = phen_env_mod3)
cat("\n", "Model 3: geography, conditioned on climate", "\n")
summary(mod3_env)
```

We can now assess the impact of climate on the genetic values of $δ^{13}C$. Consider the output from the three models. Note that the environmental variables were standardized, which means that we can directly compare their effects from the slope estimates. 

**Question 8**:

- *Start with Model 1. Is this multiple regression model statistically significant? If so, why?* 
- *Which variables provide the largest effects (use the column labeled 'Estimate', which is the partial regression coefficient)?* 
- *Do the same sort of inspection for Model 2 & 3. What can you conclude?*

## Part 3: Compare Q_{ST} to F_{ST} 

**Motivation**: Now that we have shown that genetic variation for $δ^{13}C$ within populations is significantly greater than zero (i.e. $h^2 > 0$), that differentiation for $δ^{13}C$ is statistically greater than zero (i.e. $Q_{ST} > 0$), and that climate, and to a lesser degree geography, is correlated with $δ^{13}C$ values, we can formally test whether or not differentiation for $δ^{13}C$ is unexplainable due to neutral processes such as genetic drift and gene flow. 

The general idea is use a set of genetic markers we think primarily reflects neutral processes to estimate what $Q_{ST}$ should be without any form of natural selection operating in our system. To do that, we will use 164 SNPs sampled from gene regions that have no apparent functional connection to $δ^{13}C$. This will allow us to conclude that the differentiation we see is not just different from zero (done before), but different than expectations from a neutral model.

**Goals**: The goal for this part of the laboratory is to test the hypothesis that $Q_{ST}$ is greater than $F_{ST}$. 

```{r, echo=FALSE}
snp <- read.delim(paste0(here(),"/data/WWP_SNP_genotypes.txt"), 
                  sep = "\t", header = T)
phen <- read.delim(paste0(here(),"/data/WWP_phenotype_data.txt"), 
                   sep = "\t", header = T)
env <- read.delim(paste0(here(),"/data/WWP_environmental_data.txt"), 
                  sep = "\t", header = T)
env2 <- data.frame(matrix(nrow=nrow(env), ncol=ncol(env)))
colnames(env2) <- colnames(env)
env2[,c(1:2)] <- env[,c(1:2)]
for(i in 3:ncol(env2)) {env2[,i] <- scale(env[,i], center = T, scale = T)}

snp_reformat <- hierfstat_convert(snp = snp, ids = c(1,2))

mono <- numeric(ncol(snp_reformat))
for (i in 1:ncol(snp_reformat)) 
{
  mono[i] <- length(table(snp_reformat[,i]))
}
snp_reformat2 <- snp_reformat[,-which(mono == 1)]

colnames(snp_reformat2) <- colnames(snp)[3:ncol(snp)][-which(mono == 1)]
pop_ids <- data.frame(matrix(snp[,2], nrow=nrow(snp), ncol = 1)); colnames(pop_ids) <- c("population")
```

### a) Estimate F_{ST} from SNP data

Let’s examine differentiation for the SNP data using FST. 

```{r, echo=FALSE}
fst <- varcomp.glob(levels = pop_ids, loci = snp_reformat2, diploid = T)
fst$F
```

Here is a matrix of F-statistics. These work by using as subscripts the column title relative to the row title, so the first value on the first line is the F- statistic for population relative to total (i.e. $F_{ST}$). It is calculated based on the variance components from $overall as:

$$F_{pop,tot} = \frac{\sigma^{2}_{pop}}
{\sigma^{2}_{pop}+\sigma^{2}_{ind}+\sigma^{2}_{error}}$$

We can calculate $F_{ST}$ for each SNP and plot it against heterozygosity (left). Or to make a plot like those from Arlequin, we divide the heterozygosity values by $(1-F_{ST})$ prior to plotting them (right).

```{r, fig.show='hold', echo=FALSE}
fst_snp <- fst_persnp(vc = fst$loc, names = colnames(snp_reformat2))
het_out <- het_snp(snp=snp_reformat2, finite.cor= T, names = colnames(snp_reformat2))
plot(het_out, fst_snp)
plot(het_out/(1-fst_snp), fst_snp)
```

**Question 9**:

- Do any SNPs show an unusual $F_{ST}$ value?
- If so, what could this mean?

### c) Compare Q_{ST} to F_{ST}

Now that we have inspected overall genetic differentiation among populations, let’s test whether or not $Q_{ST} > F_{ST}$ for $δ^{13}C$. The output below lists three p-values, each corresponding to a different alternative hypothesis $H_A$:

- Lower one-tailed: $H_A: Q_{ST} < F_{ST}$  
- Upper one-tailed: $H_A: Q_{ST} > F_{ST}$ 
- Two-tailed: $H_A: Q_{ST} - F_{ST}$

```{r, echo=FALSE}
phen_mod <- phen[,-c(2,4)]
snp_reformat3 <- cbind(snp[,2], snp_reformat2); colnames(snp_reformat3)[1] <- c("population")
QstFst_out <- QstFstComp(fst.dat = snp_reformat3, qst.dat = phen_mod, numpops = 10, nsim = 10000, breeding.design = "half.sib.dam", dam.offspring.relatedness = 0.25, output = "concise")
QstFst_out[[3]]
```

**Question 10**:

- Is $Q_{ST} > F_{ST}$? 
- What does this mean biologically? 
