---
title: "Week 4: Bonus Material"
author: 
- "Andrew Eckert (worked example)"
- "Helene Wagner (vignette)"
date: "`r Sys.Date()`"
show_toc: true
output:
  knitr:::html_vignette:
    toc: yes
    fig_width: 4 
    fig_height: 3.5
vignette: >
  %\VignetteIndexEntry{Week 4: Bonus Material}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## 1. Overview of Bonus Material

### a) Goals 

This bonus material shows how to:

- Perform and assess output from basic association analyses linking genetic variation with environmental variation. We will return to the topic of outlier locus detection and gene-environment associations later in the course (Week 11).

### b) Data set 

- **WWP_SNP_genotypes.txt**: SNP genotypes for all trees sampled in the field.
- **WWP_environmental_data.txt**:Environmental data collected from each plot within each population.
- **WWP_phenotype_data.txt**: Phenotypic measurements for 5 seedlings per tree made in a common garden.

### c) Required R libraries

All required packages should have been installed already when you installed 'LandGenCourse'.

```{r message=FALSE, warning=TRUE}

#devtools::install_github("kjgilbert/QstFstComp")
#source("http://bioconductor.org/biocLite.R")
#biocLite("LEA")


require(here)
require(LEA)

source(paste0(here(),"/data/supplemental_R_functions.R"))

if(!dir.exists(paste0(here(),"/data"))) dir.create(paste0(here(),"/data"))
```

## 2. Test association between genetic and environmental data

**Motivation**: So far we have shown the following: $δ^{13}C$ is genetically determined, genetically structured among populations, correlated to a variety of environmental variables, and that the structuring of genetic diversity for $δ^{13}C$ is statistically greater than that for random SNP markers. That sounds awesome, but we can go a bit further. Although we assumed that the SNP markers had no effect on $δ^{13}C$ and were neutral, we should probably check to see if that is true. To do so, we will use an environmental association approach that also corrects for background levels of population structure. Alternatively, if we had other markers that were good candidates to affect $δ^{13}C$ we could see if they were consistent with acting as the genetic architecture of $δ^{13}C$, which appears to contribute to local adaptation of P. monticola in the Lake Tahoe Basin.

**Goals & Background**: The goal of this part of the laboratory is to explore environmental associations of the SNP genotypes with the environmental data. We will do this with the relatively new R package named LEA, which allows use of latent factor mixed models (LFMM) to carry out the analysis (see more at: https://www.bioconductor.org/packages/release/bioc/html/LEA.html).

### a) Reformat SNP and environmental data

We have to reformat our data yet again. Do you not love software designers and their sadomasochistic need to make us do this? Okay, rant over. Let’s get to work.

We need to put the genotype data in a format labeled as lfmm for LEA. This format is a matrix, with sampled trees as rows and SNPs as columns. In each cell of this matrix is a 0, 1, or 2, which represents the count of a reference allele for that sampled tree for that SNP. Our reference allele will be the minor allele (i.e. the one with the lowest sample frequency across all sampled trees). I have provided a function to do this in the "supplemental_R_functions.R" file: geno_reformat(). This function operates on the snp_reformat2 object from our previous work:

```{r}
geno_snp <- geno_reformat(snp = snp_reformat2)
write.table(t(geno_snp), file = paste0(here(), "/data/geno_format.lfmm"), sep = "\t", row.names = F, col.names = F)
```

Now, we need to do the same for the environmental data. Let’s use the centered and scaled data from before, which are located in the env2 object:

```{r}
env3 <- env2[,-c(1:4)]
write.table(env3, file = paste0(here(), "/data/lfmm_env.env"), sep = "\t", row.names = F, col.names = F)
```

### b) Basic analysis
See also: https://rdrr.io/bioc/LEA/f/inst/doc/LEA.pdf

We should now be ready to use the 'lfmm' function in 'LEA'. Let’s use the following command to run a simple analysis (note that the paths here assume that the needed files are in your working directory):



```{r}
files <- c(paste0(here(), c("/data/geno_format.lfmm", "/data/lfmm_env.env")))
lfmm_out <- lfmm(input.file = files[1], 
                 environment.file = files[2], 
                 K = 5, project = "new", missing.data = T, all = T, 
                 iterations = 10000, burnin = 5000)
lfmm_out

# NEW following tutorial
project = snmf(files[1], K=1:10, entropy = TRUE, repetitions = 10,
     project = "new")
plot(project, lwd = 5, col = "red", pch=1)

project = NULL
project = lfmm(files[1], files[2], K = 3, iterations = 6000, burnin = 3000,
               repetitions = 1, project = "new")

```

### c) Choose k

You can now spend some time varying parameters. I would start with K and vary it from 2 to 10 to see how the results change. 

To get an idea of which K might be best, you can do the following. Use the pca_out <- pca(input.file="geno_format.lfmm", scale = T, center = T) and tracy.widom(<output from pca>) functions from LEA to determine the number of PCs that best explain your data. 

```{r}
pca_out <- pca(input.file = paste0(here(),"/data/geno_format.lfmm"), scale = T, center = T)
tw <- tracy.widom(pca_out)
tw
plot(tw$percentage)
```

If you do this, look in the output of tracy.widom() for the column labeled pvalues. Select K as the number of PCs with this value below a threshold such as P = 0.05. 

### d) Interpret p-values

Some of the main results you should inspect are the p-values for association of a SNP with an environmental gradient. These can be extracted for each environmental variable using the mlog10p.values() function: e.g. mlog10p.values(lfmm_out, K = 5, d = 1) will extract the –log10 p-values for the first environmental variable (higher values means lower p-values from testing the null hypothesis of no association between genotypic variation and environmental variation). 

```{r, fig.show='hold'}
# Note: 'mlog10p.values' no longer available? Use 'lfmm.pvalues' instead?
#mlog10p.values(lfmm_out, K = 5, d = 1)
P.values <- lfmm.pvalues(lfmm_out, genomic.control=TRUE, K = 5, d = 1)
P.values
plot(P.value$pvalues)
plot(sort(P.value$pvalues))
```

Questions:

- Which environmental variables have the highest –log10 p-values? 
- If you correct for multiple tests per environmental variable using a Bonferroni correction, the –log10 p-value threshold is: -log10(0.05/160) = 3.50515. Which environmental variables have SNPs with –log10 p-values greater than this threshold? 
- How do these relate to the partial regression coefficients from the multiple regression model linking genetic values of $δ^{13}C$? 
- Are the SNPs with high –log10 p-values those with $F_{ST}$ larger than the multilocus values?




